Dullahan Partial Refactor: They Work Again Edition

GoldenAlpharex opened the pull request:
## About The Pull Request
So, a few months ago I was like "hmm there's something weird going on with party pods...", which got me looking into important_recursive_hearers or something like that. I spoke about it in the coding channel and Kyler actually fixed it before I did. But I also caught a similar glitch with Dullahans, so I decided to investigate...

*Two months later...*

I present to you a partial unfuckening of the Dullahans, in that I made them fully functional once again:
 - They only hear speech through their head (not sounds, sadly, someone else would have to tell me how to do that because I otherwise really wouldn't know how to do it in a sane way), they speak through their head, runechat-included.
 - When you spawn a Dullahan, you're set to look through the Dullahan's eyes (so from their head), and that doesn't reset when you log off and back in, or admin-ghost and come back in your body.
 - When you're looking through your head, your view will no longer be reset to your body upon entering a locker, which is nice to avoid not being blind while looking through your body.
 - Dullahan heads no longer look completely lifeless and without organs. They have eyes that don't look dead and that even match the player's intended eye color.
 - Dullahan can now properly examine things from their head, which was intended and 100% not functional.
 - Dullahan heads now speak with the proper name of their owner, instead of having a random name attached to it at round-start.
 - Dullahan heads are also now properly named too.
 - Dullahans can now properly whisper, sing and do all these funny things that they were unable to do before.
 - Dullahan whispers will now properly respect the range of the whisper.
 - Dullahans can now succumb in hardcrit by whispering, as intended. This potentially fixes other species that worked similarly not being able to succumb, like abductors, although I didn't test if they normally could, I just know they absolutely will be able to now.
 - When switching from Dullahans to a different species, your old head will no longer stay behind.

I also added a proc for species to do some code when we get a ckey login in our mob, which could potentially be useful for other stuff in the future, but it was necessary here as the view is reliant on the client, which we want to ensure doesn't get weird view glitches like having their head's vision overlay while actually being centered on their body.

I also made it so `say()` now takes a `range` argument, which is `7` by default, just so things that aren't humans can also whisper and do all those kinds of things. Going with that, there's probably a few more things that will be able to be done better thanks to this, although I haven't tested every edge case with this, but I doubt it will make *much* of a difference in the future.

Closes #32236
Closes #41277
Closes #41529
Closes #54685
Closes #62411
Closes #62441

There's a few issues I wasn't able to fix during this, or that I found out while making this PR, such as dullahans not having hair when they go back from being Monkeyfied, or the fact that skillchips fail to implant in Dullahans (which, to be fair, kinda makes sense most of the time, a possible fix would be a work-around when they're forced to take an implant, like when getting equiped with an outfit that has a skillchip, but not letting it happen otherwise, unless they're holding their brain in the machine or something, maybe), but I didn't deem them nearly important enough to make the species unplayable (like it currently is), and it's barely a minor annoyance at this point.

And, no, I don't know what happened to my branch, maybe I had it for a long time and just updated it or whatever, oh well.

## Why It's Good For The Game

That way, Dullahans are not a runtime-inducing buggy mess whenever Halloween happens, and they can *actually* be used.

I was just honestly ashamed to have something so utterly broken in our code, so I fixed it out of spite.

## Changelog

Lemon told me he made longer changelogs in the past so blame this on him if you think it's too long.

:cl: GoldenAlpharex
refactor: Refactored Dullahans in many areas to fix many, many of their bugs and glitches and ensure an actually functional experience playing as one, while also improving as much code as I could along the way.
fix: Dullahans only hear speech through their head (not sounds, sadly, someone else would have to tell me how to do that because I otherwise really wouldn't know how to do it in a sane way), they speak through their head, runechat-included.
fix: When you spawn a Dullahan, you're set to look through the Dullahan's eyes (so from their head), and that doesn't reset when you log off and back in, or admin-ghost and come back in your body.
fix: When you're looking through your head, your view will no longer be reset to your body upon entering a locker, which is nice to avoid not being blind while looking through your body.
fix: Dullahan heads no longer look completely lifeless and without organs. They have eyes that don't look dead and that even match the player's intended eye color.
fix: Dullahan can now properly examine things from their head's view.
fix: Dullahan heads now speak with the proper name of their owner, instead of having a random name attached to it at round-start.
fix: Dullahan heads are also now properly named too.
fix: Dullahans can now properly whisper, sing and do all these funny things that they were unable to do before.
fix; Dullahan whispers will now properly respect the range of the whisper.
fix: Dullahans can now succumb in hardcrit by whispering, as intended. This potentially fixes other species that worked similarly not being able to succumb, like abductors.
fix: When switching from Dullahans to a different species, your old head will no longer stay behind.
/:cl:

actions-user committed:
Automatic changelog compile [ci skip]

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

actions-user committed:
Automatic changelog compile [ci skip]

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation

GoldenAlpharex committed:
We're getting somewhere

GoldenAlpharex committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into dullahan-fixes

GoldenAlpharex committed:
Updating a few variables and preventing a hard del

GoldenAlpharex committed:
Nearly functional

GoldenAlpharex committed:
Fixes double-hearing, whoops

GoldenAlpharex committed:
Properly fixes the naming of the head and its voice and removes debug code

tralezab commented:
link all the issues this fixes please

SuperNovaa41 commented:
https://github.com/tgstation/tgstation/issues?q=is%3Aissue+is%3Aopen+dullahan+
i think these are all the current dullahan issues (hopefully this fixes all of them :))))

MrMelbert commented on code/modules/mob/living/carbon/human/species_types/dullahan.dm:
```diff
@@ -20,57 +20,89 @@
+		detached_head.name = human.real_name
+		detached_head.brain.name = "[human.name]'s brain"
+
+	var/obj/item/bodypart/head/head2 = human.get_bodypart(BODY_ZONE_HEAD)
```
Can you change this var name while you're here? `head2` tells nothing

GoldenAlpharex committed:
I fixed so much and yet there's still more to fix, in the future

But hey. It works, it's more than functional, and even more functional than it's ever been before! Yay!

GoldenAlpharex committed:
I forgot to remove my debug code (again)

GoldenAlpharex committed:
Hopefully this should help with hard dels related to still being in a list of hearing_contents

tralezab commented:
i'm just gonna send it, if dullahans brokey after this any more than they brokey now you will be dragged back kicking and screaming into dullahan.dm

tralezab commented:
seems impossible to break dullahans any more though
